js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx

```diff
@@ -1,118 +1,181 @@
         }
       }
-    })();
-  }, [lru, version, cacheBreaker]);
-
-  const fetch = React.useCallback(async () => {
-    setLoading(true);
-    if (fetchState[key]) {
-      return await new Promise<ApolloQueryResult<TQuery>>((res) => {
-        fetchState[key]?.onFetched.push((value) => {
-          setCacheBreaker((v) => v + 1);
-          res(value);
+
+      const currentState = fetchState[key];
+      // Handle concurrent fetch requests
+      if (currentState) {
+        return new Promise((resolve) => {
+          currentState!.onFetched.push(resolve as any);
         });
+      }
+
+      const state = {onFetched: [] as ((value: any) => void)[]};
+      fetchState[key] = state;
+
+      const queryResult = await client.query<TQuery, TVariables>({
+        query,
+        variables,
+        fetchPolicy: 'no-cache',
       });
-    }
-    fetchState[key] = {onFetched: []};
-    // Use client.query here so that we initially use the apollo cache if any data is available in it
-    // and so that we don't subscribe to any updates to that cache (useLazyQuery and useQuery would both subscribe to updates to the
-    // cache which can be very slow)
-    const queryResult = await client.query<TQuery, TVariables>({
-      query,
-      variables,
-      fetchPolicy: 'no-cache', // Don't store the result in the cache,
-      // should help avoid page stuttering due to granular updates to the data
-    });
-    const {data} = queryResult;
-    setLoading(false);
-    lru.set(
-      'cache',
-      {data, version},
-      {
-        expiry: new Date('3000-01-01'), // never expire,
-      },
-    );
-    delete fetchState[key];
-    const onFetched = fetchState[key]?.onFetched;
-    try {
-      setData(data);
-    } catch (e) {
-      setTimeout(() => {
-        throw e;
-      });
-    }
-    onFetched?.forEach((cb) => {
-      try {
-        cb(queryResult);
-      } catch (e) {
-        setTimeout(() => {
-          throw e;
-        });
+
+      const {data} = queryResult;
+      await cacheManager.set(data, version);
+
+      const onFetchedHandlers = state.onFetched;
+      if (fetchState[key] === state) {
+        delete fetchState[key]; // Clean up fetch state after handling
       }
-    });
 
-    return queryResult;
-  }, [client, key, lru, query, variables, version]);
+      onFetchedHandlers.forEach((handler) => handler(data)); // Notify all waiting fetches
+
+      return data;
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
+  );
+}
+
+export function useGetCachedData() {
+  const {getCacheManager} = useContext(IndexedDBCacheContext);
 
+  return useCallback(
+    async <TQuery,>({key, version}: {key: string; version: number}) => {
+      const cacheManager = getCacheManager<TQuery>(key);
+      return await cacheManager.get(version);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
+  );
+}
+
+const contextValue = createIndexedDBCacheContextValue();
```

Comment:
Hmm if the value of the context is this constant, would there be harm in making it a global and skipping all the React bits?  Maybe there are places where it's useful to override the cache manager in a component subtree?

 I will read on!

Reply:
This was mostly for tests, but I think I don't need it. I was trying to figure out how state was leaking between test cases and figured out its actually due to jest not clearing out allTicks between tests (I had to manually call `jest.runAllTicks()` at the end of each test to resolve this. Jest.clearAllTimers() doesn't clear ticks (eg. Promises/Async functions).